### PR TITLE
[RFR] Default value function

### LINKED ISSUE #274: Feature Request: defaultValue prop in SimpleForm should support dynamic value
My first try with Global Default Value

```
const generateCode = {
    number: 1,
    call: function() {
        let code =  'CODE' + this.number;
        this.number++;
        return code;
    }
};

const productDefaultValue = { created_at: new Date(), code:  generateCode.call()};

export const ProductCreate = (props) => (
    <Create {...props}>
        <SimpleForm defaultValue={productDefaultValue}>
            <TextInput source="code" />
            <DateInput source="created_at" />
        </SimpleForm>
    </Create>
);
```

The product 'code' is expected as 'CODE1', 'CODE2' and so on everytime we press CREATE button.
But it does not, it's always 'CODE1'.
'created_at' prop has the same story.

Then, I try Per Field Default Value
```
export const ProductCreate = (props) => (
    <Create {...props}>
        <SimpleForm>
            <TextInput source="code" defaultValue={generateCode.call()}/>
            <DateInput source="created_at" defaultValue={new Date()}/>
        </SimpleForm>
    </Create>
);
```
The result is good for 'created_at' prop, it always receives new date value.
But the 'code' prop is not pretty good.
It is CODE1 at 1st time, CODE4 at 2nd time, CODE8 at 3rd time.

Seem that 'generateCode.call()' gets called everytime React renders as below
`_react2.default.createElement(_input.TextInput, _extends({}, MA_DH, { defaultValue: generateCode.call()})), ...);`

I think we should somehow support dynamic default value.

If there is better way, please light me up.
Thank you very much.

### PR DESCRIPTION
**Pull request by fzaninotto**
Allow `defaultValue` to be a function for form and input components.

Use case:

```js
export const PostCreate = (props) => (
    <Create {...props}>
        <SimpleForm>
            <DisabledInput source="id" defaultValue={() => uuid()}/>
            ...
        </SimpleForm>
    </Create>
);
```

Closes #274

### CODE REVIEW

File: src/mui/form/SimpleForm.js [hunk 1]
```diff
@@ -40,8 +44,11 @@ const ReduxForm = reduxForm({
     validate: validateForm,
 })(SimpleForm);
 
+const getInitialValues = once((children, record, defaultValue) =>
```
Comment (djhi) [inline src/mui/form/SimpleForm.js]: Shouldn't this be extracted to `../form/getInitialValues` ? I see you use it in TabbedForm too

Comment (fzaninotto) [inline src/mui/form/SimpleForm.js]: you're absolutely right

**Comment by fzaninotto** [comment]
Back to WIP: using `once()` makes the function execute only once for the entire application, while I want it to execute once each time the component is mounted.

**Comment by fzaninotto** [comment]
I'll have to use `reselect` instead of `once`

**Comment by fzaninotto** [comment]
And back to RFR

File: src/CrudRoute.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { IndexRoute, Route } from 'react-router';
 import { createRoutesFromReactChildren } from 'react-router/lib/RouteUtils';
+import { pure } from 'recompose';
```
Comment (djhi) [inline src/CrudRoute.js]:
```js
import pure from 'recompose/pure';
````
?

Comment (djhi) [inline src/CrudRoute.js]: Will be better when using webpack 2 and dead code elimination

**Comment by fzaninotto** [comment]
Code review applied

**Pull request merged**